service/dap/server.go

```diff
@@ -1196,94 +1200,128 @@ func (s *Session) onSetBreakpointsRequest(request *dap.SetBreakpointsRequest) {
+	response.Body.Breakpoints = breakpoints
+
+	s.send(response)
+}
+
+type bpMetadata struct {
+	name         string
+	condition    string
+	hitCondition string
+	logMessage   string
+}
+
+type bpLocation struct {
+	File  string
+	Line  int
+	Addr  uint64
+	Addrs []uint64
+}
+
+// setBreakpoints is a helper function for setting source, function and instruction
+// breakpoints. It takes the prefix of the name for all breakpoints that should be
+// included, the total number of breakpoints, and functions for computing the metadata
+// and the location. The location is computed separately because this may be more
+// expensive to compute and may not always be necessary.
+func (s *Session) setBreakpoints(prefix string, totalBps int, metadataFunc func(i int) *bpMetadata, locFunc func(i int) (*bpLocation, error)) []dap.Breakpoint {
 	// See https://github.com/golang/vscode-go/issues/163 for details.
 	// If a breakpoint:
 	// -- exists and not in request => ClearBreakpoint
 	// -- exists and in request => AmendBreakpoint
 	// -- doesn't exist and in request => SetBreakpoint
 
-	// Get all existing breakpoints that match for this source.
-	sourceRequestPrefix := fmt.Sprintf("sourceBp Path=%q ", request.Arguments.Source.Path)
-	existingBps := s.getMatchingBreakpoints(sourceRequestPrefix)
+	// Get all existing breakpoints matching the prefix.
+	existingBps := s.getMatchingBreakpoints(prefix)
 	bpAdded := make(map[string]struct{}, len(existingBps))
+	for _, bp := range existingBps {
+		existingBps[bp.Name] = bp
+	}
 
-	// Amend existing breakpoints.
-	breakpoints := make([]dap.Breakpoint, len(request.Arguments.Breakpoints))
-	for i, want := range request.Arguments.Breakpoints {
-		reqString := fmt.Sprintf("%s Line=%d Column=%d", sourceRequestPrefix, want.Line, want.Column)
-		var err error
-		got, ok := existingBps[reqString]
-		if !ok {
+	breakpoints := make([]dap.Breakpoint, totalBps)
+	for i := 0; i < totalBps; i++ {
+		want := metadataFunc(i)
+		got, ok := existingBps[want.name]
+		if got == nil || !ok {
 			// Skip if the breakpoint does not already exist.
-			// These will be created after deleting existing
-			// breakpoints to avoid conflicts.
 			continue
 		}
-		if _, ok := bpAdded[reqString]; ok {
-			err = fmt.Errorf("breakpoint exists at %q, line: %d, column: %d", request.Arguments.Source.Path, want.Line, want.Column)
+
+		var err error
+		if _, ok := bpAdded[want.name]; ok {
+			err = fmt.Errorf("breakpoint already exists")
 		} else {
-			got.Cond = want.Condition
-			got.HitCond = want.HitCondition
-			got.Tracepoint = want.LogMessage != ""
-			got.UserData = want.LogMessage
+			got.Cond = want.condition
+			got.HitCond = want.hitCondition
+			got.Tracepoint = want.logMessage != ""
+			got.UserData = want.logMessage
 			err = s.debugger.AmendBreakpoint(got)
-			bpAdded[reqString] = struct{}{}
 		}
-
-		updateBreakpointsResponse(breakpoints, i, err, got, clientPath)
+		bpAdded[want.name] = struct{}{}
```

Comment:
Why do this use empty struct here as opposed to bool?

Reply:
maps of empty structs use less memory than bools when implementing a set. If we want to change it for readability, these sets are never that large so probably have minimal space and time savings.